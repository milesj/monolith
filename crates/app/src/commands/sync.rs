use super::syncs::codeowners::SyncCodeownersArgs;
use super::syncs::hooks::SyncHooksArgs;
use crate::session::CliSession;
use clap::Subcommand;
use starbase::AppResult;
use starbase_styles::color;
use tracing::warn;

#[derive(Clone, Debug, Subcommand)]
pub enum SyncCommands {
    #[command(
        name = "codeowners",
        about = "Aggregate and sync code owners to a `CODEOWNERS` file."
    )]
    Codeowners(SyncCodeownersArgs),

    #[command(
        name = "hooks",
        about = "Generate and sync hook scripts for the workspace configured VCS."
    )]
    Hooks(SyncHooksArgs),

    #[command(
        name = "projects",
        about = "Sync all projects and configs in the workspace."
    )]
    Projects,
}

pub async fn sync(session: CliSession) -> AppResult {
    warn!(
        "This command is deprecated. Use {} instead.",
        color::shell("moon sync projects")
    );

    crate::commands::syncs::projects::sync(session).await
}
