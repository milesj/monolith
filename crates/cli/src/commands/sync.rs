use crate::helpers::{build_dep_graph, create_progress_bar, load_workspace};
use moon_runner::Runner;

pub async fn sync() -> Result<(), Box<dyn std::error::Error>> {
    let done = create_progress_bar("Syncing projects...");

    let mut workspace = load_workspace().await?;
    let project_graph = workspace.generate_project_graph().await?;
    let mut project_count = 0;
    let mut dep_builder = build_dep_graph(&workspace, &project_graph);

    for project in project_graph.get_all()? {
        dep_builder.sync_project(project)?;
        project_count += 1;
    }

    let dep_graph = dep_builder.build();
    let mut runner = Runner::new(workspace);
    let results = runner.run(dep_graph, None).await?;

    if runner.has_failed() {
        done("Failed to sync projects", false);
    } else {
        done(
            format!("Successfully synced {} projects", project_count).as_ref(),
            true,
        );
    }

    runner.render_results(&results)?;

    Ok(())
}
