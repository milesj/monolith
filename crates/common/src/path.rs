pub use relative_path::*;
use starbase_styles::color;
use std::path::Path;

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[cfg(unix)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

pub enum RelativeFrom<'path> {
    Project(&'path str),
    Workspace,
}

pub fn expand_to_workspace_relative<P: AsRef<str>>(
    from_format: RelativeFrom,
    path: P,
) -> WorkspaceRelativePathBuf {
    let path = standardize_separators(path.as_ref());

    match from_format {
        RelativeFrom::Project(source) => {
            // Root-level project
            if source.is_empty() || source == "." {
                WorkspaceRelativePathBuf::from(path)

                // Project-level, prefix with source path
            } else {
                let project_source = standardize_separators(source);

                if let Some(negated_glob) = path.strip_prefix('!') {
                    WorkspaceRelativePathBuf::from(format!("!{project_source}")).join(negated_glob)
                } else {
                    WorkspaceRelativePathBuf::from(project_source).join(path)
                }
            }
        }
        RelativeFrom::Workspace => WorkspaceRelativePathBuf::from(path),
    }
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let path = path.as_ref();

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            color::path(path)
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    #[cfg(windows)]
    {
        format!("{}.exe", name.as_ref())
    }

    #[cfg(not(windows))]
    {
        name.as_ref().into()
    }
}
