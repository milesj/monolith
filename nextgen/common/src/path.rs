pub use relative_path::{RelativePath, RelativePathBuf};

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}
