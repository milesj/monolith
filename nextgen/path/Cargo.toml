[package]
name = "moon_path"
version = "0.1.0"
edition = "2021"

[dependencies]
relative-path = { version = "1.8.0", features = ["serde"] }
serde = { workspace = true }
