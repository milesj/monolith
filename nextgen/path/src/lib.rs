pub use relative_path::{RelativePath, RelativePathBuf};

// pub enum PathType {
//     File(String),
//     Glob(String),
// }

// pub enum Location<T> {
//     Absolute(T),
//     ProjectRelative(T),
//     WorkspaceRelative(T),
// }

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[inline]
pub fn expand_to_workspace_relative<F, P>(file: F, project_source: P) -> WorkspaceRelativePathBuf
where
    F: AsRef<str>,
    P: AsRef<str>,
{
    let file = file.as_ref();
    let project_source = project_source.as_ref();

    // Workspace relative negative glob
    if file.starts_with("/!") || file.starts_with("!/") {
        return WorkspaceRelativePathBuf::from(format!("!{}", standardize_separators(&file[2..])))
            .normalize();
    }

    // Workspace relative file/glob
    if let Some(path) = file.strip_prefix('/') {
        return WorkspaceRelativePathBuf::from(standardize_separators(path)).normalize();
    }

    // Project relative negative glob
    if let Some(glob) = file.strip_prefix('!') {
        return WorkspaceRelativePathBuf::from(format!("!{}", project_source))
            .join(standardize_separators(glob))
            .normalize();
    }

    // Project relative file/glob
    WorkspaceRelativePathBuf::from(project_source)
        .join(standardize_separators(file))
        .normalize()
}
