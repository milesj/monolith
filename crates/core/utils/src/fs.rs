use async_recursion::async_recursion;
use futures::future::try_join_all;
use json_comments::StripComments;
use moon_error::{map_io_to_fs_error, map_json_to_error, map_yaml_to_error, MoonError};
use regex::Regex;
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime};
use tokio::fs;

#[inline]
pub fn clean_json<T: AsRef<str>>(json: T) -> Result<String, MoonError> {
    let json = json.as_ref();

    // Remove comments
    let mut stripped = String::with_capacity(json.len());

    StripComments::new(json.as_bytes())
        .read_to_string(&mut stripped)
        .map_err(MoonError::Unknown)?;

    // Remove trailing commas
    let stripped = Regex::new(r",(?P<valid>\s*})")
        .unwrap()
        .replace_all(&stripped, "$valid");

    Ok(String::from(stripped))
}

#[inline]
pub async fn copy_file<S: AsRef<Path>, D: AsRef<Path>>(from: S, to: D) -> Result<(), MoonError> {
    let from = from.as_ref();
    let to = to.as_ref();
    let to_dir = to.parent().unwrap();

    create_dir_all(to_dir).await?;

    fs::copy(from, to)
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[async_recursion]
pub async fn copy_dir_all<T: AsRef<Path> + Send>(
    from_root: T,
    from: T,
    to_root: T,
) -> Result<(), MoonError> {
    let from_root = from_root.as_ref();
    let from = from.as_ref();
    let to_root = to_root.as_ref();
    let entries = read_dir(from).await?;
    let mut files = vec![];
    let mut dirs = vec![];

    for entry in entries {
        let path = entry.path();

        if path.is_file() {
            files.push(copy_file(
                path.to_owned(),
                to_root.join(path.strip_prefix(from_root).unwrap()),
            ));
        } else {
            dirs.push(path);
        }
    }

    // Copy files before dirs incase an error occurs
    try_join_all(files).await?;

    // Copy dirs in sequence for the same reason
    for dir in dirs {
        copy_dir_all(from_root, &dir, to_root).await?;
    }

    Ok(())
}

#[inline]
pub async fn create_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if !path.exists() {
        fs::create_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub fn find_upwards<F, P>(name: F, dir: P) -> Option<PathBuf>
where
    F: AsRef<str>,
    P: AsRef<Path>,
{
    let dir = dir.as_ref();
    let findable = dir.join(name.as_ref());

    if findable.exists() {
        return Some(findable);
    }

    match dir.parent() {
        Some(parent_dir) => find_upwards(name, parent_dir),
        None => None,
    }
}

#[inline]
pub async fn metadata<T: AsRef<Path>>(path: T) -> Result<std::fs::Metadata, MoonError> {
    let path = path.as_ref();

    fs::metadata(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))
}

#[inline]
pub async fn read_dir<T: AsRef<Path>>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        results.push(entry);
    }

    Ok(results)
}

#[async_recursion]
pub async fn read_dir_all<T: AsRef<Path> + Send>(path: T) -> Result<Vec<fs::DirEntry>, MoonError> {
    let path = path.as_ref();
    let handle_error = |e| map_io_to_fs_error(e, path.to_path_buf());

    let mut entries = fs::read_dir(path).await.map_err(handle_error)?;
    let mut results = vec![];

    while let Some(entry) = entries.next_entry().await.map_err(handle_error)? {
        if let Ok(file_type) = entry.file_type().await {
            if file_type.is_dir() {
                results.extend(read_dir_all(&entry.path()).await?);
            } else {
                results.push(entry);
            }
        }
    }

    Ok(results)
}

#[inline]
pub async fn read<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    let path = path.as_ref();
    let data = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(data)
}

#[inline]
pub async fn read_json<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = read_json_string(path).await?;

    let json: D =
        serde_json::from_str(&contents).map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
    clean_json(read(path).await?)
}

#[inline]
pub async fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
where
    P: AsRef<Path>,
    D: DeserializeOwned,
{
    let path = path.as_ref();
    let contents = fs::read_to_string(path)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    let json: D =
        serde_yaml::from_str(&contents).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    Ok(json)
}

#[inline]
pub async fn remove<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.is_file() {
        remove_file(path).await?;
    } else if path.is_dir() {
        remove_dir_all(path).await?;
    }

    Ok(())
}

#[inline]
pub async fn remove_file<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_file(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

#[inline]
pub async fn remove_dir_all<T: AsRef<Path>>(path: T) -> Result<(), MoonError> {
    let path = path.as_ref();

    if path.exists() {
        fs::remove_dir_all(&path)
            .await
            .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;
    }

    Ok(())
}

pub type RemoveDirContentsResult = (usize, u64);

pub async fn remove_dir_stale_contents<P: AsRef<Path>>(
    dir: P,
    duration: Duration,
) -> Result<RemoveDirContentsResult, MoonError> {
    let mut files_deleted: usize = 0;
    let mut bytes_saved: u64 = 0;
    let threshold = SystemTime::now() - duration;

    for entry in read_dir(dir.as_ref()).await? {
        let path = entry.path();

        if path.is_file() {
            let mut bytes = 0;

            if let Ok(metadata) = entry.metadata().await {
                bytes = metadata.len();

                if let Ok(filetime) = metadata.accessed().or_else(|_| metadata.created()) {
                    if filetime > threshold {
                        // Not stale yet
                        continue;
                    }
                } else {
                    // Not supported in environment
                    continue;
                }
            }

            if remove_file(path).await.is_ok() {
                files_deleted += 1;
                bytes_saved += bytes;
            }
        }
    }

    Ok((files_deleted, bytes_saved))
}

#[inline]
pub async fn rename<F: AsRef<Path>, T: AsRef<Path>>(from: F, to: T) -> Result<(), MoonError> {
    let from = from.as_ref();

    fs::rename(from, to.as_ref())
        .await
        .map_err(|e| map_io_to_fs_error(e, from.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write<T: AsRef<Path>>(path: T, data: impl AsRef<[u8]>) -> Result<(), MoonError> {
    let path = path.as_ref();

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_json<P, D>(path: P, json: &D, pretty: bool) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data = if pretty {
        serde_json::to_string_pretty(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    } else {
        serde_json::to_string(&json).map_err(|e| map_json_to_error(e, path.to_path_buf()))?
    };

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

#[inline]
pub async fn write_yaml<P, D>(path: P, yaml: &D) -> Result<(), MoonError>
where
    P: AsRef<Path>,
    D: ?Sized + Serialize,
{
    let path = path.as_ref();
    let data =
        serde_yaml::to_string(&yaml).map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

    fs::write(path, data)
        .await
        .map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

    Ok(())
}

pub mod temp {
    use super::*;
    use moon_constants::CONFIG_DIRNAME;
    use std::fs; // TEMPORARILY sync

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache/temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{:x}.{}", md5::compute(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> Result<Option<String>, MoonError> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        // Temp files only last for 4 hours (half a workday)
        let threshold = SystemTime::now() - Duration::from_secs(60 * 60 * 4);

        if let Ok(metadata) = file.metadata() {
            if let Ok(filetime) = metadata.created() {
                if filetime > threshold {
                    fs::remove_file(file)?;

                    return Ok(None);
                }
            }
        }

        Ok(Some(fs::read_to_string(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> Result<(), MoonError> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write(path, data.as_ref())?;

        Ok(())
    }
}

pub mod sync {
    use super::*;
    use std::fs::read_to_string;

    #[inline]
    pub fn read<P>(path: P) -> Result<String, MoonError>
    where
        P: AsRef<Path>,
    {
        let path = path.as_ref();
        let data = read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        Ok(data)
    }

    #[inline]
    pub fn read_json<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents = read_json_string(path)?;

        let json: D = serde_json::from_str(&contents)
            .map_err(|e| map_json_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }

    #[inline]
    pub fn read_json_string<T: AsRef<Path>>(path: T) -> Result<String, MoonError> {
        clean_json(read(path.as_ref())?)
    }

    #[inline]
    pub fn read_yaml<P, D>(path: P) -> Result<D, MoonError>
    where
        P: AsRef<Path>,
        D: DeserializeOwned,
    {
        let path = path.as_ref();
        let contents =
            read_to_string(path).map_err(|e| map_io_to_fs_error(e, path.to_path_buf()))?;

        let json: D = serde_yaml::from_str(&contents)
            .map_err(|e| map_yaml_to_error(e, path.to_path_buf()))?;

        Ok(json)
    }
}
