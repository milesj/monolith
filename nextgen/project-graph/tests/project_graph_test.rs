use moon_common::Id;
use moon_config::{
    DependencyConfig, DependencyScope, InheritedTasksManager, NodeConfig, ToolchainConfig,
    WorkspaceConfig, WorkspaceProjects,
};
use moon_project_builder::DetectLanguageEvent;
use moon_project_graph2::{
    ExtendProjectData, ExtendProjectEvent, ExtendProjectGraphData, ExtendProjectGraphEvent,
    ProjectGraph, ProjectGraphBuilder, ProjectGraphBuilderContext,
};
use moon_task_builder::DetectPlatformEvent;
use rustc_hash::FxHashMap;
use starbase_events::{Emitter, EventState};
use starbase_sandbox::{assert_snapshot, create_sandbox};
use starbase_utils::fs;
use starbase_utils::string_vec;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::sync::RwLock;

#[derive(Default)]
struct GraphContainer {
    inherited_tasks: InheritedTasksManager,
    toolchain_config: ToolchainConfig,
    workspace_config: WorkspaceConfig,
    workspace_root: PathBuf,
}

impl GraphContainer {
    pub fn new(root: &Path) -> Self {
        let mut graph = Self {
            workspace_root: root.to_path_buf(),
            ..Default::default()
        };
        graph.toolchain_config.node = Some(NodeConfig::default());
        graph.workspace_config.projects = WorkspaceProjects::Globs(string_vec!["*"]);
        graph
    }

    pub fn create_context(&self) -> ProjectGraphBuilderContext {
        ProjectGraphBuilderContext {
            extend_project: Emitter::<ExtendProjectEvent>::new(),
            extend_project_graph: Emitter::<ExtendProjectGraphEvent>::new(),
            detect_language: Emitter::<DetectLanguageEvent>::new(),
            detect_platform: Emitter::<DetectPlatformEvent>::new(),
            inherited_tasks: &self.inherited_tasks,
            toolchain_config: &self.toolchain_config,
            vcs: None,
            workspace_config: &self.workspace_config,
            workspace_root: &self.workspace_root,
        }
    }

    pub async fn build_graph<'l>(&self, context: ProjectGraphBuilderContext<'l>) -> ProjectGraph {
        let mut builder = ProjectGraphBuilder::new(context).await.unwrap();
        builder.load_all().await.unwrap();

        let graph = builder.build().await.unwrap();
        graph.get_all().unwrap();
        graph
    }
}

async fn generate_project_graph(fixture: &str) -> ProjectGraph {
    let sandbox = create_sandbox(fixture);
    let container = GraphContainer::new(sandbox.path());
    let context = container.create_context();

    container.build_graph(context).await
}

fn map_ids(ids: Vec<&Id>) -> Vec<String> {
    ids.iter().map(|id| id.to_string()).collect()
}

mod project_graph {
    use super::*;

    mod dependencies {
        use super::*;

        #[tokio::test]
        async fn explicit_depends_on() {
            let graph = generate_project_graph("dependencies").await;

            assert_snapshot!(graph.to_dot());
        }

        #[tokio::test]
        async fn lists_ids_of_dependencies() {
            let graph = generate_project_graph("dependencies").await;

            assert_eq!(
                map_ids(graph.dependencies_of(&graph.get("a").unwrap()).unwrap()),
                ["b"]
            );
            assert_eq!(
                map_ids(graph.dependencies_of(&graph.get("b").unwrap()).unwrap()),
                ["c"]
            );
            assert_eq!(
                map_ids(graph.dependencies_of(&graph.get("c").unwrap()).unwrap()),
                string_vec![]
            );
            assert_eq!(
                map_ids(graph.dependencies_of(&graph.get("d").unwrap()).unwrap()),
                ["c", "b", "a"]
            );
        }

        #[tokio::test]
        async fn lists_ids_of_dependents() {
            let graph = generate_project_graph("dependencies").await;

            assert_eq!(
                map_ids(graph.dependents_of(&graph.get("a").unwrap()).unwrap()),
                ["d"]
            );
            assert_eq!(
                map_ids(graph.dependents_of(&graph.get("b").unwrap()).unwrap()),
                ["d", "a"]
            );
            assert_eq!(
                map_ids(graph.dependents_of(&graph.get("c").unwrap()).unwrap()),
                ["d", "b"]
            );
            assert_eq!(
                map_ids(graph.dependents_of(&graph.get("d").unwrap()).unwrap()),
                string_vec![]
            );
        }
    }

    mod aliases {
        use super::*;

        async fn aliases_project_graph() -> ProjectGraph {
            let sandbox = create_sandbox("aliases");
            let container = GraphContainer::new(sandbox.path());
            let context = container.create_context();

            // Set aliases for projects
            context
                .extend_project_graph
                .on(
                    |event: Arc<ExtendProjectGraphEvent>,
                     data: Arc<RwLock<ExtendProjectGraphData>>| async move {
                        let mut data = data.write().await;

                        for (id, source) in &event.sources {
                            let alias_path = source.join("alias").to_path(&event.workspace_root);

                            if alias_path.exists() {
                                data.aliases.insert(
                                    fs::read_file(alias_path).unwrap().trim().to_owned(),
                                    id.to_owned(),
                                );
                            }
                        }

                        Ok(EventState::Continue)
                    },
                )
                .await;

            // Set implicit deps for projects
            context
                .extend_project
                .on(
                    |event: Arc<ExtendProjectEvent>,
                     data: Arc<RwLock<ExtendProjectData>>| async move {
                        let mut data = data.write().await;

                        if event.project_id == "explicit-and-implicit" || event.project_id == "implicit" {
                            data.dependencies.push(DependencyConfig {
                                id: "@three".into(),
                                scope: DependencyScope::Build,
                                ..Default::default()
                            });
                        }

                        if event.project_id == "implicit" {
                            data.dependencies.push(DependencyConfig {
                                id: "@one".into(),
                                scope: DependencyScope::Peer,
                                ..Default::default()
                            });
                        }

                        Ok(EventState::Continue)
                    },
                )
                .await;

            container.build_graph(context).await
        }

        #[tokio::test]
        async fn loads_aliases() {
            let graph = aliases_project_graph().await;

            assert_snapshot!(graph.to_dot());

            assert_eq!(
                graph
                    .nodes
                    .into_iter()
                    .map(|(id, node)| (id, node.alias))
                    .collect::<FxHashMap<_, _>>(),
                FxHashMap::from_iter([
                    ("alias-one".into(), Some("@one".to_owned())),
                    ("alias-two".into(), Some("@two".to_owned())),
                    ("alias-three".into(), Some("@three".to_owned())),
                    ("explicit".into(), None),
                    ("explicit-and-implicit".into(), None),
                    ("implicit".into(), None),
                ])
            );
        }

        #[tokio::test]
        async fn can_get_projects_by_alias() {
            let graph = aliases_project_graph().await;

            assert!(graph.get("@one").is_ok());
            assert!(graph.get("@two").is_ok());
            assert!(graph.get("@three").is_ok());

            assert_eq!(graph.get("@one").unwrap(), graph.get("alias-one").unwrap());
            assert_eq!(graph.get("@two").unwrap(), graph.get("alias-two").unwrap());
            assert_eq!(
                graph.get("@three").unwrap(),
                graph.get("alias-three").unwrap()
            );
        }

        #[tokio::test]
        async fn can_depends_on_by_alias() {
            let graph = aliases_project_graph().await;

            assert_eq!(
                map_ids(
                    graph
                        .dependencies_of(&graph.get("explicit").unwrap())
                        .unwrap()
                ),
                ["alias-one", "alias-two"]
            );

            assert_eq!(
                map_ids(
                    graph
                        .dependencies_of(&graph.get("explicit-and-implicit").unwrap())
                        .unwrap()
                ),
                ["alias-three", "alias-two"]
            );

            assert_eq!(
                map_ids(
                    graph
                        .dependencies_of(&graph.get("implicit").unwrap())
                        .unwrap()
                ),
                ["alias-three", "alias-one"]
            );
        }
    }
}
