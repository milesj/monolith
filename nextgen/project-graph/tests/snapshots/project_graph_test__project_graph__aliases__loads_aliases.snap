---
source: nextgen/project-graph/tests/project_graph_test.rs
expression: graph.to_dot()
---
digraph {
    0 [ label="tasks" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    1 [ label="alias-one" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    2 [ label="alias-three" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    3 [ label="implicit" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    4 [ label="dupes-task-deps" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    5 [ label="alias-two" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    6 [ label="dupes-depends-on" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    7 [ label="explicit-and-implicit" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    8 [ label="explicit" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    0 -> 1 [ label="peer" arrowhead=none]
    0 -> 2 [ label="peer" arrowhead=none]
    3 -> 1 [ label="peer" arrowhead=box, arrowtail=box]
    3 -> 2 [ label="build" arrowhead=box, arrowtail=box]
    0 -> 3 [ label="peer" arrowhead=none]
    4 -> 1 [ label="peer" arrowhead=box, arrowtail=box]
    4 -> 1 [ label="peer" arrowhead=box, arrowtail=box]
    6 -> 5 [ label="development" arrowhead=box, arrowtail=box]
    6 -> 5 [ label="build" arrowhead=box, arrowtail=box]
    7 -> 5 [ label="development" arrowhead=box, arrowtail=box]
    7 -> 2 [ label="build" arrowhead=box, arrowtail=box]
    8 -> 5 [ label="production" arrowhead=box, arrowtail=box]
    8 -> 1 [ label="development" arrowhead=box, arrowtail=box]
}

